hazelcast/proxy/reliable_topic.py

```diff
@@ -1,12 +1,331 @@
+    is also known as a publish/subscribe (pub/sub) messaging model. Publish and subscriptions are cluster-wide. When a
+    member subscribes for a topic, it is actually registering for messages published by any member in the cluster,
+    including the new members joined after you added the listener.
+
+    Messages are ordered, meaning that listeners(subscribers) will process the messages in the order they are actually
+    published.
+
+    Hazelcast's Reliable Topic uses the same Topic interface as a regular topic. The main difference is that Reliable
+    Topic is backed up by the Ringbuffer data structure, a replicated but not partitioned data structure that stores
+    its data in a ring-like structure.
+    """
+
+    def __init__(self, client, service_name, name):
+        super(ReliableTopic, self).__init__(client, service_name, name)
+
+        config = client.config.reliable_topic_configs.get(name, None)
+        if config is None:
+            config = ReliableTopicConfig()
+
+        self.client = client
+        self.config = config
+        self._topic_overload_policy = self.config.topic_overload_policy
+        self.ringbuffer = client.get_ringbuffer("_hz_rb_" + name)
+        self._message_listeners_lock = threading.RLock()
+        self._message_listeners = {}
+
+    def add_listener(self, reliable_topic_listener):
+        """
+        Subscribes to this reliable topic. When someone publishes a message on this topic, on_message() method of
+        ReliableTopicListener is called.
+
+        :param ReliableTopicListener: (Class), class to be used when a message is published.
+        :return: (str), a registration id which is used as a key to remove the listener.
+        """
+        if not isinstance(reliable_topic_listener, ReliableMessageListener):
+            raise IllegalArgumentError("Message listener is not an instance of ReliableTopicListener")
+
+        registration_id = str(uuid4())
+        listener = _MessageListener(registration_id, self, self._to_object, reliable_topic_listener)
+        with self._message_listeners_lock:
+            self._message_listeners[registration_id] = listener
+
+        listener.start()
+        return registration_id
+
+    def _add_with_backoff(self, item):
+        sleep_time = _INITIAL_BACKOFF
+        while True:
+            seq_id = self.ringbuffer.add(item, overflow_policy=OVERFLOW_POLICY_FAIL).result()
+            if seq_id != -1:
+                return
+            time.sleep(sleep_time)
+            sleep_time *= 2
+            if sleep_time > _MAX_BACKOFF:
+                sleep_time = _MAX_BACKOFF
+
+    def _add_or_fail(self, item):
+        seq_id = self.ringbuffer.add(item, overflow_policy=OVERFLOW_POLICY_FAIL).result()
+        if seq_id == -1:
+            raise TopicOverflowError("failed to publish message to topic: " + self.name)
+
     def publish(self, message):
-        raise NotImplementedError
+        """
```

Comment:
This is a blocking API. In general the API provided in the 3.x versions are non-blocking apart from the proxy creations and listener registrations. So, I think `publish` should be non-blocking too.